### DISCUSSION TITLE
Can a MenuItem populated from an ItemsControl be made to look the same as directly populating with MenuItems?

### DISCUSSION THREAD

**Discussion by wgraaf** [AvaloniaUI/Avalonia#14918] [Q&A] [answered]
I have a MenuItem that I want to populate with a collection of sub-menu items, each with an icon and a header.
If I directly place items in the parent menu item the icon appears on the left, followed by the header.

```
        <MenuItem Header="header1" Command="{Binding MenuCommand1}">
          <MenuItem.Icon>
            <Image Source="/Assets/checkMark.bmp" />
          </MenuItem.Icon>
        </MenuItem>
        <MenuItem Header="header1" Command="{Binding MenuCommand1}">
          <MenuItem.Icon>
            <Image Source="/Assets/checkMark.bmp" />
          </MenuItem.Icon>
        </MenuItem>

```
If I populate from an items control, the menu item leaves an empty space to the left and only uses the space for the header.

```
        <ItemsControl ItemsSource="{Binding ColumnVisibilities}">
          <ItemsControl.ItemTemplate>
            <DataTemplate>
              <MenuItem Header="{Binding HeaderText}" Command="{Binding [MenuCommand}">](url)
                <MenuItem.Icon>
                  <Image Source="/Assets/checkMark.bmp" />
                </MenuItem.Icon>
              </MenuItem>
            </DataTemplate>
          </ItemsControl.ItemTemplate>
        </ItemsControl>

```
![image](https://github.com/AvaloniaUI/Avalonia/assets/123667061/a6f5a5a5-21ca-436a-b969-2dfb29c68bd1)

Is there a way to make the menu item look like a normal menu item when populated from an items control?

**Comment by maxkatz6** [comment]
I don't see well nearly enough from this code, but I assume you are putting ItemsControl inside of the single MenuItem. Which adds this offset - these are menu items inside of the menu item.

**Comment by BAndysc** [comment, marked as answer]
MenuItem itself is an ItemsControl, there is no need to use an additional ItemsControl. Just bind ItemsSource in MenuItem and use styles to bind the properties:
```
<Menu ItemsSource="{Binding ColumnVisibilities}">
    <Menu.Styles>
        <Style Selector="MenuItem">
            <! -- <Setter Property="ItemsSource" Value="{Binding NestedMenu}" /> -->
            <Setter Property="Header" Value="{Binding Header}" />
            <Setter Property="Command" Value="{Binding MenuCommand}"  />
          </Style>
    </Menu.Styles>
</Menu>
```

**Reply by wgraaf** [reply]
That works, thanks.
This is what I ended up with:
```
      <MenuItem Header="View" ItemsSource="{Binding ColumnVisibilities}">
        <MenuItem.Styles>
          <Style Selector="MenuItem">
            <Setter Property="Header" Value="{Binding HeaderText}"/>
            <Setter Property="Command" Value="{Binding ToggleColumnCommand}"/>
            <Setter Property="StaysOpenOnClick" Value="True"/>
            <Setter Property="Icon">
              <Template>
                <Image Source="/Assets/checkMark.bmp" IsVisible="{Binding IsVisible}"/>
              </Template>
            </Setter>
          </Style>
        </MenuItem.Styles>
      </MenuItem>
```
Am I right in assuming that using the MenuItem ItemSource precludes the possibility of adding other items to the menu?
(i.e. if ItemsSource is used, ALL items in the (sub)menu come from there, and adding other menu items will not work)

**Reply by timunie** [reply]
How is your ItemsSource defined? If it is an ObservableCollection, adding and removing items should work.
